### DISCUSSION TITLE
Storing Stats and Modified Stats

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2624] [Evennia Questions] [answered]
So, this is probably more a Python question than an Evennia question, but as I think it's likely that others may have the same question in the future, I'll ask it here.

I plan for my characters to have stats/attributes/whatever term you like, such as strength, dexterity, hitroll, saves, etc. At least with a few of these (strength, dexterity, intelligence, wisdom and constitution, notably), it is going to be valuable for the player to know both the unmodified-by-equipment-and-effects number ("base") and the modified number ("modified"). I had originally planned on storing all of the various stats (those with base and modified values, and those for which the base of 0 is irrelevant) in a dictionary, so that I can iterate through all of the stats easily. I had also planned to be able to iterate through the modified stats. My question is, what is the way to store both the base and modified data so that it is a) iterable and b) convenient to reference?

The two options I thought of were:

1. one unified dictionary, containing a key string ("strength") and a value list (17, 22) consisting of the base value and the modified value; or
2. two separate dictionaries, one for base values and one for modified values.

The first seems like a pain to access, as you'll have to remember (and know, if you are not me and reading the code) that item [0] of the list is base, and [1] is modified. 

The second will work, but just seems vaguely inelegant to me. I know I should probably just pick one and do it, since both will work, but I'm trying to learn the "right" way to do this stuff to the extent that such a way exists.

Thanks in advance,

Random

**Comment by ChrisLR** [comment]
Consider if you need to access the base and modified values at the same time.
The first option is much better if you do, while the second one is easier if you don't.
In both cases, consider using properties or methods to access your stats rather than accessing them directly.
It will make it easier to change how things work later on.

**Reply by RandomEC** [reply]
I am almost certainly going to write methods to interact with the values. The only time I can think of that you would interact with both at (approximately) the same time is in simply displaying them both to the player through "score" or similar. Almost everything in the game is going to interact with modified.

I guess a third option is just to store base strength (for example), and then call a method on the character to get modified strength, which method iterates through all the equipment being worn and adds up all the modifiers. That is obviously more computationally intensive than either of the above options, but is much more flexible.

**Reply by Tegiminis** [reply]
Using a property method on the typeclass to return the modified value is what I do. Ideally, you don't store the modified value at all; rather, you store the base, any modifiers, and compute on-demand what the modified value is. This ensures the modified value is always concurrent. I wouldn't worry about it being computationally intensive.

**Comment by jaborsh** [comment, marked as answer]
While it is presently only in the `develop` branch of Evennia, you could certainly recreate the process in the `master` branch.

Whitenoise and Griatch spent a lot of time creating the traits contrib, which allows for a properly organized, managed, and accessible method of doing exactly what you described. I'll link it here.

https://github.com/evennia/evennia/tree/develop/evennia/contrib/rpg/traits

**Reply by RandomEC** [reply]
I suspect that I will end up doing something similar to this and Tegiminis' answer, above. I SUPER don't understand properties right now, and don't want to bolt-on a property based solution without understanding them better. Likely I will write my own, using the traits contrib as a guide, just so that I can learn for myself how properties work.

Thanks!

**Reply by gtaylor** [reply]
Whoops, didn't mean to unmark and remark the answer in this thread. Was reading along on GitHub mobile and thought the button did something else!
